codegen/builtin_fs.go

```diff
@@ -0,0 +1,499 @@
+package codegen
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/openllb/hlb/parser"
+)
+
+type Scratch struct{}
+
+func (s Scratch) Call(ctx context.Context, ret Value, opts Options) error {
+	return ret.Set(llb.Scratch())
+}
+
+type Image struct{}
+
+func (i Image) Call(ctx context.Context, ret Value, opts Options, ref string) error {
```

Comment:
Could the `Call()` func belong to a `Callable` interface? Perhaps taking `args []Value` instead of splatting the args into the function signature

Reply:
I started off with a `Callable` interface, but I realized it was prone to error and code duplication where every function needed to validate the length of `args []Value` and then handle each type coercion.

```go
func (i Image) Call(ctx context.Context, ret Value, args []Value, opts Options) error {
    if len(args) != 1 {
        return errors.WithStack(ErrCodeGen{ProgramCounter(ctx), ErrNumArgs})
    }

    ref, err := args[0].AsString()
    if err != nil {
        return err
    }
```

When I refactored to splatting args into signature already typed, it reduced the complexity of all the builtin implementations, and I found a few `args[<index>]` and length checks were wrong.

In addition, I wanted to support things like `t time.Time` in the signature to ensure that all times in codegen are parsed with the same RFC format string.